### DISCUSSION TITLE
Are relative paths for mounting functioning?

### DISCUSSION THREAD

**Discussion by stephaje** [moghtech/komodo#180] [Q&A] [answered]
I am having a difficult time getting relative path volumes working without issues.
I have:

- A pre-existing "stack" directory with several stack subfolders (e.x. `~/stacks`) and 0770 permissions
- Each stack subfolder with a compose.yaml and directories for mounting (e.x. `~/stacks/container`)
- The compose.yaml mounts are all relative, (e.x. `./data:/data`, where `./data` is at `~/stacks/container/data`)
- My stacks directory mounted in komodo
- A Komodo stack created and configured to use `Files On Server`, which sees the existing compose.yaml file

However, any time I try to start the stack through Komodo, the container has permission errors for accessing the relative path volume. I don't have this issue when using `docker compose up` in the container directory manually.

Am I missing a setup step here?

**Comment by mbecker20** [comment]
Are you running Periphery in Docker container? In this case docker compose get very confused about the relative paths when running in container, it pretty much will only work with absolute paths when running Periphery in Docker container.

Should work as you expect with `systemd` deploy option for Periphery, instead of container: https://komo.do/docs/connect-servers#install-the-periphery-agent---systemd

**Comment by stephaje** [comment]
I am using a docker container. I'd really like to get this to work, so I'm going to dig into it more.

Problem: relative paths are interpreted by the periphery container as paths within the /etc/komodo/stacks/* directories, but since the process is on the host, the path needs to be relative to the host directories. Confirmed that using relative paths in the periphery container puts the volumes in /etc/komodo/stacks on the host machine.

Compose volume: `./data:/config`
Up via host: `"/home/me/web-server/chowdown/data:/config:rw"`
Up via komodo container: `"/etc/komodo/stacks/chowdown/data:/config:rw"`

Proposed workaround: Create a link between /etc/komodo/stacks and /home/me/web-server on the host machine.
`sudo ln -s /home/me/web-server /etc/komodo; sudo mv /etc/komodo/web-server /etc/komodo/stacks`
**Success!** Maybe not the cleanest solution, but this works.

I think, in theory, you could get this working inversely from the periphery container, i.e. read the stack bind, set up a sym link from the host directory structure to the default structure inside the periphery container, and then run the compose up command from the host directory structure instead. I'm sure there are a lot of caveats here, but I hope it's something that is considered in the future.

**Reply by mbecker20** [reply]
What happens if you add mount to Periphery container, something like: `/home/me/web-server:/etc/komodo`?

**Reply by stephaje** [reply]
I'm already using `~/web-server:/etc/komodo/stacks`, but even after changing to `/home/me/web-server:/etc/komodo/stacks`, the container is still set up with `/etc/komodo/stacks/chowdown/data:/config:rw`. As long as the working directory for the `up` command is `/etc/komodo/stacks/chowdown`, relative paths will always resolve there.

Now, if I make a `/home/me/web-server -> /etc/komodo/stacks` symlink in the periphery container, and then navigate to `/home/me/web-server/chowdown` and run `docker compose up -d` there, the mount in the periphery container matches my local filesystem, so it works.

Surprisingly, just setting the run directory in the Komodo UI for the stack to `/home/me/web-server/chowdown` does not lead to the container being set up correctly. Using `docker container inspect`, the output says that the working directory is the original `/etc/komodo/stacks/*`:
```
"Labels": {
    ...
    "com.docker.compose.project": "chowdown",
    "com.docker.compose.project.config_files": "/etc/komodo/stacks/chowdown/compose.yaml",
    "com.docker.compose.project.working_dir": "/etc/komodo/stacks/chowdown",
    "com.docker.compose.service": "chowdown",
    "com.docker.compose.version": "2.29.7"
}
```

It seems like the run directory setting doesn't behave like I would expect it to. I would expect that if it was left blank, it would run the following:
`root:/etc/komodo/stacks/chowdown # docker compose up -d ...`
and if it were set to a non default value, e.x. `/home/me/web-server/chowdown`, it would run:
`root:/home/me/web-server/chowdown # docker compose up -d ...` which properly sets the  working_dir label to `"com.docker.compose.project.working_dir": "/home/hal/web-server/chowdown"`,
but that doesn't appear to be the case.

**Reply by mbecker20** [reply, marked as answer]
Interesting. You can change the configuration for the stack directory, in periphery container env vars. What if you try this instead:

`PERIPHERY_STACK_DIR=/home/me/web-server` (ref: https://github.com/mbecker20/komodo/blob/main/config/periphery.config.toml#L27)

and change the mount as well:

` - /home/me/web-server:/home/me/web-server`

This should cause the paths to be the same outside and inside the container, basically tricking docker compose without having to do symlink.

**Reply by stephaje** [reply]
That works! Confirmed that the working_dir label is appropriately set with this configuration. Thanks for helping to find a more graceful solution.

**Reply by icf20** [reply]
this is not working properly nor a proper solution, the repos remain available only in the container

**Reply by GerardPolloRebozado** [reply]
that fixed the issue, thank you

**Comment by cvalcke958** [comment]
I equally struggle with the location for compose files. I figured out from this thread to create a directory under /etc/komodo/stacks (default) and put the compose.yaml there. I tried a simple /etc/komodo/stacks/web with the nginx compose file but upon deploy, I get:
ERROR: Failed to read compose file contents at "/etc/komodo/stacks/web/compose.yaml"
Q: did it find the file but couldn't read?  or is there anything else that I need to "mount" ?  (running in LXC container on Proxmox)

**Reply by mbecker20** [reply]
Yes, you should mount `- /etc/komodo:/etc/komodo` to the Periphery container.

**Reply by cvalcke958** [reply]
apologies, i'm docker newbie. where/how do I specify the mount?    
looking at the komodo yaml file in /opt/komodo, it seems to have the stacks entry already:

>      ## manage stack files in a docker volume,
>       ## or change it to an accessible host directory.
>       - stacks:/etc/komodo/stacks
>       ## Optionally mount a path to store compose files
>       # - /path/to/compose:/host/compose

do I add another to that file as below (or somewhere else)?
        - /etc/komodo:/etc/komodo

**Reply by mbecker20** [reply]
You replace `- stacks:/etc/komodo/stacks` with `- /etc/komodo:/etc/komodo` (/etc/komodo/stacks is already contained within /etc/komodo). Then you will end up with folders accessible on your host, outside docker, at `/etc/komodo/stacks`, `/etc/komodo/ssl`, and so on.

I definitely get why many people were confused here, this distribution of compose files has issues. I've fixed them in the upcoming release, you can already use them instead as your starting point (they are backward compatible / already usable before I release 1.17). You can find them here: https://github.com/mbecker20/komodo/blob/1.17.0/compose/mongo.compose.yaml#L63 and https://github.com/mbecker20/komodo/blob/1.17.0/compose/compose.env#L122.

If you have other linux experience outside of Docker, you may prefer to run Periphery with `systemd`. It actually makes this simpler. I wrote more on this here: #220 . In this case, you can remove the `periphery` service from the default docker compose file you used to launch Komodo, and instead use the install script: https://github.com/mbecker20/komodo/blob/main/scripts/readme.md.

**Reply by cvalcke958** [reply]
Got It! new compose files help to understand. Thx for suggestion to use systemd, great option that I missed in my first pass through the setup.

**Reply by cvalcke958** [reply]
Up and running on two servers with systemd peripheries !! Beautiful & thx for great homelab app :)

**Comment by ovizii** [comment]
Sorry to reply to this closed thread, but I am trying to puzzle together the answers from this thread as the docs are pretty silent on this subject. I tried reading and understanding https://komo.do/docs/docker-compose, but I am failing to put it all together. 

Please give me a hand. Lets say on my host, I have this fiel structure:

```
/docker
/docker/komodo/docker-compose.yml
/docker/stack01/docker-compose.yml
```

and so on. 

**Questions:**

- which paths do I need to mount via volumes for core/periphery?
- what do I set these variables to: PERIPHERY_REPO_DIR, PERIPHERY_STACK_DIR and PERIPHERY_ROOT_DIRECTORY

**Reply by stephaje** [reply]
Here is my periphery compose.yml file, where my stacks are as follows:
```
/home/me/web-server/komodo/compose.yml
/home/me/web-server/stack01/compose.yml
```

```
services:
  periphery:
    image: ghcr.io/mbecker20/periphery:latest
    labels:
      komodo.skip: # Prevent Komodo from stopping with StopAllContainers
    restart: unless-stopped
    logging:
      driver: local
    ports:
      - 8120:8120
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /proc:/proc
      - ./ssl-certs:/etc/komodo/ssl
      - ./repos:/etc/komodo/repos
      - /home/me/web-server:/home/me/web-server
    environment:
      PERIPHERY_PASSKEYS: <key>
      PERIPHERY_ALLOWED_IPS:
      PERIPHERY_SSL_ENABLED: true
      PERIPHERY_INCLUDE_DISK_MOUNTS: /etc/hostname
      PERIPHERY_STACK_DIR: /home/me/web-server
```

I think the equivalent in your case would be
```
volumes:
  - /docker:/docker
environment:
  PERIPHERY_STACK_DIR: /docker
```
  , and similar for the repo and root directories I would assume, but I can't speak from experience.

**Reply by ovizii** [reply]
Thanks, I guess that will do the trick but then I noticed you didn't mention: `PERIPHERY_ROOT_DIRECTORY` is that not important? What's supposed to go in there?
